lib/core/checker/include/type_expression_checker.h

```diff
@@ -0,0 +1,97 @@
+
+#pragma once
+
+#include "core/ast/ast.h"
+#include "core/ast/expression.h"
+#include "core/checker/common.h"
+#include "core/interning/interned_string.h"
+
+#include <deque>
+#include <string>
+#include <unordered_map>
+#include <vector>
+
+namespace dbuf::checker {
+
+class TypeExpressionChecker {
+  template <typename T>
+  void operator()(const ast::TypeExpression &, const T &) {}
+
+  explicit TypeExpressionChecker(ast::AST &ast);
+
+  void operator()();
+
+  void operator()(const ast::TypeExpression &type_expression);
+
+  void operator()(const ast::TypeExpression &expected_type, const ast::VarAccess &expression);
+
+  void operator()(const ast::Expression &expected, const ast::Expression &got) {
+    // TODO(SphericalPotatoInVacuum): compare expressions
+  }
+
+  void operator()(const ast::TypeExpression &expected_type, const ast::TypeExpression &expression);
+
+  void operator()(const ast::TypeExpression &expected_type, const ast::ConstructedValue &value);
+
+  template <typename T>
+  void operator()(const ast::TypeExpression &expected_type, const ast::ScalarValue<T> &value) {
+    if (expected_type.identifier.name != GetTypename(value)) {
+      errors_.emplace_back(Error {
+          .message = "Got value of type \"" + GetTypename(value).GetString() +
+                     "\", but expected type is \"" + expected_type.identifier.name.GetString() +
+                     "\""});
+      return;
+    }
+  }
+
+  void operator()(
+      const std::vector<ast::TypedVariable> &type_dependencies,
+      const ast::TypeExpression &type_expression);
+
+  void operator()(const std::unordered_map<InternedString, ast::Message> &messages);
+  void operator()(const std::unordered_map<InternedString, ast::Enum> &enums);
```

Comment:
Why do we process maps instead of tree nodes?

Reply:
A node is a singular object, map just encapsulates a number of children in one container, but the children are still those objects inside the container and note the container itself. It just helps manage them.